quickwit/quickwit-control-plane/src/indexing_plan.rs

```diff
@@ -0,0 +1,657 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::cmp::Ordering;
+use std::collections::HashMap;
+
+use itertools::Itertools;
+use quickwit_cluster::ClusterMember;
+use quickwit_config::{SourceConfig, CLI_INGEST_SOURCE_ID, INGEST_API_SOURCE_ID};
+use quickwit_proto::indexing_api::IndexingTask;
+use serde::Serialize;
+
+/// A [`PhysicalIndexingPlan`] defines the list of indexing tasks
+/// each indexer, identified by its node ID, should run.
+/// TODO(fmassot): a metastore version number will be attached to the plan
+/// to identify if the plan is up to date with the metastore.
+#[derive(Debug, PartialEq, Clone, Serialize)]
+pub struct PhysicalIndexingPlan {
+    indexing_tasks_per_node_id: HashMap<String, Vec<IndexingTask>>,
+}
+
+impl PhysicalIndexingPlan {
+    /// Creates a [`PhysicalIndexingPlan`] from a non empty list of node IDs.
+    pub fn new(node_ids: Vec<String>) -> Self {
+        assert!(!node_ids.is_empty());
+        let indexing_tasks_per_indexer = node_ids
+            .into_iter()
+            .map(|node_id| (node_id, Vec::new()))
+            .collect();
+        Self {
+            indexing_tasks_per_node_id: indexing_tasks_per_indexer,
+        }
+    }
+
+    /// Adds the indexing task to the list of tasks assigned to the given node ID.
+    pub fn assign_indexing_task(&mut self, node_id: &String, indexing_task: IndexingTask) {
+        let node_indexing_tasks = self
+            .indexing_tasks_per_node_id
+            .get_mut(node_id)
+            .expect("The node must always be present.");
```

Comment:
this method has a hidden contract and does not need to be public (if node_id is not present it panics)